scubainit/src/shadow.rs

```diff
@@ -0,0 +1,100 @@
+use crate::entfiles::{EntFileReader, EntFileWriter, Entry};
+use crate::util::{maybe_parse, to_string_or_empty};
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct ShadowEntry {
+    pub name: String,
+    pub passwd: String,
+    pub last_change_date: Option<u32>,
+    pub min_password_age: Option<u32>,
+    pub max_password_age: Option<u32>,
+    pub warn_period: Option<u32>,
+    pub inact_period: Option<u32>,
+    pub expire_date: Option<u32>,
+    // reserved
+}
+
+pub type ShadowFileReader<'a> = EntFileReader<'a, ShadowEntry>;
+pub type ShadowFileWriter<'a> = EntFileWriter<'a, ShadowEntry>;
+
+impl Entry for ShadowEntry {
+    fn from_line(line: &str) -> Option<ShadowEntry> {
+        // https://man7.org/linux/man-pages/man5/shadow.5.html
+        let mut parts = line.split(":");
+
+        // Any integer parsing errors result in the entire entry being skipped via the final .ok()?
+        Some(ShadowEntry {
+            name: parts.next()?.to_string(),
+            passwd: parts.next()?.to_string(),
+            last_change_date: maybe_parse(parts.next()?).ok()?,
+            min_password_age: maybe_parse(parts.next()?).ok()?,
+            max_password_age: maybe_parse(parts.next()?).ok()?,
+            warn_period: maybe_parse(parts.next()?).ok()?,
+            inact_period: maybe_parse(parts.next()?).ok()?,
+            expire_date: maybe_parse(parts.next()?).ok()?,
```

Comment:
Thanks for the commentary on idiomatic patterns and the specific example. I was unaware of `Result::transpose()` and `Option::transpose()` which made keeping these at one line nearly impossible.

I do think there is a balance to be achieved between writing idiomatic code that leverages the std lib, and avoiding excessive code duplication.  IMO, if a pattern is copy/pasted across more than a few lines, there is an opportunity for a helper function.

---

This comment actually reminded me of a fairly big deficiency in this version of implementation:
- IO errors (on file read) would result in iteration silently stopping.
- Any parsing errors (missing fields or invalid integers) would result in the line simply being skipped.

This is a little too forgiving; I'd rather bail loudly, rather than ignoring a problematic line (and for scubainit,  subsequently duplicating it).

So commit `scubainit: Change EntFileReader to yield Result<Entry>` changes this to return a `Result` from the iterator. This matches [`std::fs::ReadDir`](https://doc.rust-lang.org/std/fs/struct.ReadDir.html).

Handling these errors and coalescing them into a single type resulted in even *more* code duplication (`map_err()`), so I doubled down on the custom code and implemented `EntLineParser` with common methods used across all three implementations.

PTAL and let me know what you think.

Reply:
It looks a lot cleaner now.